### PR TITLE
[AutoDiff upstream] Add derivative function SILDeclRefs.

### PR DESCRIPTION
**Pull request by dan-zheng**
`@differentiable` attribute on protocol requirements and non-final class members
will produce derivative function entries in witness tables and vtables.

This patch adds an optional derivative function configuration
(`AutoDiffDerivativeFunctionIdentifier`) to `SILDeclRef` to represent these
derivative function entries.

Derivative function configurations consist of:
- A derivative function kind (JVP or VJP).
- Differentiability parameter indices.

```
sil-decl-ref ::= '#' sil-identifier ('.' sil-identifier)* sil-decl-subref?
sil-decl-subref ::= '!' sil-decl-subref-part ('.' sil-decl-lang)? ('.' sil-decl-autodiff)?
sil-decl-autodiff ::= sil-decl-autodiff-kind '.' sil-decl-autodiff-indices
ail-decl-autodiff-kind ::= 'jvp' | 'vjp'
sil-decl-autodiff-indices ::= [SU]+
                               ^~ 'S' is set, 'U' is unset
```

Resolves TF-1209.
Enables TF-1212: upstream derivative function entries in witness tables/vtables.

---

Derivative function `SILDeclRef` examples:
```swift
protocol Protocol {
  @differentiable(wrt: (x, y))
  func f(_ x: Float, _ y: Float) -> Float
}

sil hidden @witness_method : $@convention(thin) <T where T : Protocol> (@in T) -> () {
bb0(%0 : $*T):
  // Original function reference.
  %1 = witness_method $T, #Protocol.f : <Self where Self : Protocol> (Self) -> (Float, Float) -> Float : $@convention(witness_method: Protocol) <τ_0_0 where τ_0_0 : Protocol> (@in_guaranteed τ_0_0) -> (Float, Float) -> Float

  // Derivative function references.
  %2 = witness_method $T, #Protocol.f!jvp.SSS : <Self where Self : Protocol> (Self) -> (Float, Float) -> Float : $@convention(witness_method: Protocol) <τ_0_0 where τ_0_0 : Protocol> (@in_guaranteed τ_0_0) -> (Float, Float) -> Float
  %3 = witness_method $T, #Protocol.f!jvp.UUS : <Self where Self : Protocol> (Self) -> (Float, Float) -> Float : $@convention(witness_method: Protocol) <τ_0_0 where τ_0_0 : Protocol> (@in_guaranteed τ_0_0) -> (Float, Float) -> Float
  %4 = witness_method $T, #Protocol.f!vjp.SSS : <Self where Self : Protocol> (Self) -> (Float, Float) -> Float : $@convention(witness_method: Protocol) <τ_0_0 where τ_0_0 : Protocol> (@in_guaranteed τ_0_0) -> (Float, Float) -> Float
  %5 = witness_method $T, #Protocol.f!vjp.UUS : <Self where Self : Protocol> (Self) -> (Float, Float) -> Float : $@convention(witness_method: Protocol) <τ_0_0 where τ_0_0 : Protocol> (@in_guaranteed τ_0_0) -> (Float, Float) -> Float

  %6 = tuple ()
  return %6 : $()
}
```

### CODE REVIEW

**Comment by dan-zheng** [comment]
@slavapestov: would you like to review this patch, as a SIL code owner?
You previously approved this patch in 2018 on `tensorflow` branch: https://github.com/apple/swift/pull/21224

**Comment by dan-zheng** [comment]
@swift-ci Please smoke test

File: include/swift/AST/AutoDiff.h
Comment (dan-zheng) [inline include/swift/AST/AutoDiff.h]: This can be simplified after https://github.com/apple/swift/pull/29105:
```suggestion
    auto derivativeCanGenSig = derivativeGenericSignature.getCanonicalSignature();
```

I'll land this in a follow-up.

**Comment by dan-zheng** [comment]
@swift-ci Please smoke test macOS platform

**Review by rxwei — APPROVED** [review]

**Comment by dan-zheng** [comment]
Both macOS and Linux CI failed mysteriously. Retriggering.
@swift-ci Please smoke test Linux

**Comment by dan-zheng** [comment]
@swift-ci Please smoke test Linux

**Comment by dan-zheng** [comment]
@swift-ci Please smoke test Linux

**Comment by dan-zheng** [comment]
@swift-ci Please test

**Comment by dan-zheng** [comment]
SwiftPM tests on Linux seem to have flaked again (for the fourth time):
<details>
```
<unknown>:0: error: module file '/home/buildnode/jenkins/workspace/swift-PR-Linux-smoke-test/branch-master/tmp_swiftpm/ModuleCache/34DIH411VVHZV/CDispatch-E28X8IBKETOX.pcm' is out of date and needs to be rebuilt: signature mismatch
10:31:25 <unknown>:0: note: imported by module 'CoreFoundation' in '/home/buildnode/jenkins/workspace/swift-PR-Linux-smoke-test/branch-master/tmp_swiftpm/ModuleCache/34DIH411VVHZV/CoreFoundation-1HECCF3FGVFVS.pcm'
10:31:25 <unknown>:0: error: missing required modules: 'CDispatch', '_SwiftDispatchOverlayShims'
10:31:25 /home/buildnode/jenkins/workspace/swift-PR-Linux-smoke-test/branch-master/swiftpm/Tests/PackageLoadingTests/PDLoadingTests.swift:50: error: PackageDescription4_2LoadingTests.testBasics : failed - 
10:31:25 Test Case 'PackageDescription4_2LoadingTests.testBasics' failed (0.778 seconds)
10:31:25 Test Suite 'PackageDescription4_2LoadingTests' failed at 2020-03-22 17:31:06.456
10:31:25 	 Executed 1 test, with 1 failure (0 unexpected) in 0.778 (0.778) seconds
10:31:25 Test Suite 'Selected tests' failed at 2020-03-22 17:31:06.456
10:31:25 	 Executed 1 test, with 1 failure (0 unexpected) in 0.778 (0.778) seconds
```
</details>
@swift-ci Please smoke test Linux

**Comment by dan-zheng** [comment]
Merging to unblock progress. Happy to address feedback later!

**Pull request merged**
